engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/TypeOfNode.java

```diff
@@ -0,0 +1,44 @@
+package org.enso.interpreter.node.expression.builtin.meta;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.nodes.Node;
+import com.oracle.truffle.api.profiles.BranchProfile;
+import org.enso.interpreter.Constants;
+import org.enso.interpreter.dsl.BuiltinMethod;
+import org.enso.interpreter.runtime.Context;
+import org.enso.interpreter.runtime.builtin.Builtins;
+import org.enso.interpreter.runtime.error.PanicException;
+import org.enso.interpreter.runtime.library.dispatch.TypesLibrary;
+
+@BuiltinMethod(
+    type = "Meta",
+    name = "type_of_builtin",
+    description = "Returns the type of a value.")
+public class TypeOfNode extends Node {
+  private @Child InteropLibrary library =
+      InteropLibrary.getFactory().createDispatched(Constants.CacheSizes.BUILTIN_INTEROP_DISPATCH);
+  private @Child TypesLibrary types =
+      TypesLibrary.getFactory().createDispatched(Constants.CacheSizes.BUILTIN_INTEROP_DISPATCH);
+
+  Object execute(Object value) {
```

Comment:
Can't we just add specializations to this to make it work for numbers?

We could also use the `AcceptsError` annotation to make it accept the dataflow error and also report the type accordingly, instead of delegating this to Enso.

It seems worthwhile to me, to make this logic centralized - some other places in the interpreter may also need to get the type of an object and it would be good if they could use the same Node - and not have to account for special handling for numbers, if possible.